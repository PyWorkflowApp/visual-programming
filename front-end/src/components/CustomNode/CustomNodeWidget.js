import React from 'react';
import * as _ from 'lodash';
import { PortWidget } from '@projectstorm/react-diagrams';
import StatusLight from '../StatusLight';
import GraphView from './GraphView';
import NodeConfig from './NodeConfig';
import '../../styles/CustomNode.css';
import * as API from '../../API';

export default class CustomNodeWidget extends React.Component {

    constructor(props) {
        super(props);
        this.state = {showConfig: false, showGraph: false};
        this.icon = '9881';
    }

    componentDidMount() {
      this.toggleConfig = this.toggleConfig.bind(this);
      this.toggleGraph = this.toggleGraph.bind(this);
      this.handleDelete = this.handleDelete.bind(this);
      this.acceptConfiguration = this.acceptConfiguration.bind(this);
    }

    // show/hide node configuration modal
    toggleConfig = () =>  {
        this.setState({showConfig: !this.state.showConfig});
    }

    // show/hide node graph modal
    toggleGraph = () =>  {
      this.setState({showGraph: !this.state.showGraph});
    }

    // delete node from diagram model and redraw diagram
    handleDelete = () =>  {
        API.deleteNode(this.props.node).then(() => {
            this.props.node.remove();
            this.props.engine.repaintCanvas();
        }).catch(err => console.log(err));
    }

    acceptConfiguration(optionsData, flowData) {
        API.updateNode(this.props.node, optionsData, flowData).then(() => {
            this.props.node.setStatus("configured");
            this.forceUpdate();
            console.log("Node updated");
            this.props.engine.repaintCanvas();
        }).catch(err => console.log(err));
    }

    render() {
        const engine = this.props.engine;
        const allPorts = _.values(this.props.node.getPorts());
        const ports = allPorts.filter(p => !p.options.name.includes("flow"));
        const flowInPort = allPorts.find(p => p.options.name === "flow-in");
        const flowOutPort = allPorts.find(p => p.options.name === "flow-out");
        // group ports by type (in/out)
        const sortedPorts = _.groupBy(ports, p => p.options.in === true ? "in" : "out");
        // create PortWidget array for each type
        const portWidgets = {};
        for (let portType in sortedPorts) {
            portWidgets[portType] = sortedPorts[portType].map(port =>
                <PortWidget engine={engine} port={port} key={port.getID()}>
                    <div className="triangle-port" />
                </PortWidget>
            );
        }

        const flowPortWidgets = [flowInPort, flowOutPort].filter(p => p).map(port =>
            <PortWidget engine={engine} port={port} key={port.getID()}
                        className={`flow-port-div flow-port-div-${port.options.in ? "in" : "out"}`}>
                <div className="flow-port" />
            </PortWidget>
        );


        let graphView;
        let width = 40;
        if (this.props.node.options.node_type !== "flow_control") {
            graphView = (
                <div className="custom-node-tabular" onClick={this.toggleGraph}>
                  <img src="tabular-icon.png" alt="Tabular" style={{width:25, height:25}}/>
                </div>
            );
            width = 80;
        }

        return (
            <div className="custom-node-wrapper">
                <div className="custom-node-name">{this.props.node.options.name}</div>
                <div className="custom-node" style={{ borderColor: this.props.node.options.color, width: width }}>
                    <div className="custom-node-icons">
                        <div className="custom-node-configure" onClick={this.toggleConfig}>
                            {String.fromCharCode(this.icon)}
                        </div>
                        <NodeConfig node={this.props.node}
                            show={this.state.showConfig}
                            toggleShow={this.toggleConfig}
                            onDelete={this.handleDelete}
                            onSubmit={this.acceptConfiguration} />
                        {graphView}
                        <GraphView node={this.props.node}
                            show={this.state.showGraph}
                            toggleShow={this.toggleGraph}
                            onDelete={this.handleDelete}
                            onSubmit={this.acceptConfiguration} />
                    </div>
                    {flowPortWidgets}
                    <div className="port-col port-col-in">
                        { portWidgets["in"] }
                    </div>
                    <div className="port-col port-col-out">
                        { portWidgets["out"] }
                    </div>
                </div>
                <StatusLight status={this.props.node.options.status} />
                <div className="custom-node-description">{this.props.node.config.description}</div>
            </div>
        );
    }
}
