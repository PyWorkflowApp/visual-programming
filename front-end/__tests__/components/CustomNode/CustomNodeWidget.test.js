import React from 'react'
import createEngine from '@projectstorm/react-diagrams';
import { render } from '@testing-library/react'
import { shallow, mount } from 'enzyme';
import CustomNodeWidget from '../../../src/components/CustomNode/CustomNodeWidget';
import CustomNodeModel from '../../../src/components/CustomNode/CustomNodeModel';

global.console = {log: jest.fn(), error: jest.fn()}


describe('Validate CustomNodeWidget', () => {

  beforeEach(() => {
    global.fetch = jest.fn(() => Promise.resolve({
      ok: true,
      data: [],
      json: jest.fn(() => []),
      text: jest.fn(() => Promise.resolve({})),
      headers:{
        get: (s)=>{
          if (s === "content-type") {
            return "text";
          }

          if (s === "Content-Disposition") {
            return "filenameToDownload";
          }
        }
      }
    }));

    const createElement = document.createElement.bind(document);
document.createElement = (tagName) => {
    if (tagName === 'canvas') {
        return {
            getContext: () => ({}),
            measureText: () => ({})
        };
    }
    return createElement(tagName);
};

  });

  it('Display CustomNodeWidget', () => {
    const node = new CustomNodeModel({id: "myId"});
    const model = {
      node: node,
      globals: {}
    };
    const engine = createEngine();
    engine.setModel(model);

    const customNodeWidget = render(
      <CustomNodeWidget  engine={engine} node={node} />
    );
    expect(customNodeWidget).toMatchSnapshot();
  });

  it('Validates CustomNodeWidget', () => {
    const node = new CustomNodeModel({id: "myId"});
    const model = {
      node: node,
      globals: {}
    };
    const engine = createEngine();
    engine.setModel(model);

    const nodeWidget = shallow(<CustomNodeWidget
        engine={engine}
        node={node}/>);

   expect(nodeWidget.state('showConfig')).toBe(false);
   nodeWidget.find({ className: 'custom-node-configure' }).simulate('click');

   expect(nodeWidget.state('showConfig')).toBe(true);
  });

  it('Mounts CustomNodeWidget', () => {
    const node = new CustomNodeModel({id: "myId"});
    const model = {
      node: node,
      globals: {}
    };
    const engine = createEngine();
    engine.setModel(model);

    const nodeWidget = mount(<CustomNodeWidget
        engine={engine}
        node={node}/>);

    expect(nodeWidget.state('showConfig')).toBe(false);
    expect(nodeWidget.state('showGraph')).toBe(false);
  });

  it('Accepts configuration', () => {
      const node = new CustomNodeModel({id: "myId"});
      const repaintCanvas = jest.fn(() => [])
      const engine = {
        repaintCanvas: repaintCanvas
      }
      const props = { node: node, engine: engine};
      const nodeWidget = new CustomNodeWidget(props);
      nodeWidget.acceptConfiguration({}, {});
  });
})
