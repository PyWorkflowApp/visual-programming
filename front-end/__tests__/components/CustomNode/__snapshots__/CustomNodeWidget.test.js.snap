// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate CustomNodeWidget Display CustomNodeWidget 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="custom-node-wrapper"
      >
        <div
          class="custom-node-name"
        />
        <div
          class="custom-node"
          style="width: 80px;"
        >
          <div
            class="custom-node-icons"
          >
            <div
              class="custom-node-configure"
            >
              ⚙
            </div>
            <div
              class="custom-node-tabular"
            >
              <img
                alt="Tabular"
                src="tabular-icon.png"
                style="width: 25px; height: 25px;"
              />
            </div>
          </div>
          <div
            class="port flow-port-div flow-port-div-in"
            data-name="flow-in"
            data-nodeid="myId"
          >
            <div
              class="flow-port"
            />
          </div>
          <div
            class="port-col port-col-in"
          >
            <div
              class="port "
              data-name="in-0"
              data-nodeid="myId"
            >
              <div
                class="triangle-port"
              />
            </div>
          </div>
          <div
            class="port-col port-col-out"
          >
            <div
              class="port "
              data-name="out-0"
              data-nodeid="myId"
            >
              <div
                class="triangle-port"
              />
            </div>
          </div>
        </div>
        <div
          class="StatusLight"
        >
          <div
            class="StatusLightItem"
            style="background-color: red;"
          />
          <div
            class="StatusLightItem"
            style="background-color: white;"
          />
          <div
            class="StatusLightItem"
            style="background-color: white;"
          />
        </div>
        <div
          class="custom-node-description"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="custom-node-wrapper"
    >
      <div
        class="custom-node-name"
      />
      <div
        class="custom-node"
        style="width: 80px;"
      >
        <div
          class="custom-node-icons"
        >
          <div
            class="custom-node-configure"
          >
            ⚙
          </div>
          <div
            class="custom-node-tabular"
          >
            <img
              alt="Tabular"
              src="tabular-icon.png"
              style="width: 25px; height: 25px;"
            />
          </div>
        </div>
        <div
          class="port flow-port-div flow-port-div-in"
          data-name="flow-in"
          data-nodeid="myId"
        >
          <div
            class="flow-port"
          />
        </div>
        <div
          class="port-col port-col-in"
        >
          <div
            class="port "
            data-name="in-0"
            data-nodeid="myId"
          >
            <div
              class="triangle-port"
            />
          </div>
        </div>
        <div
          class="port-col port-col-out"
        >
          <div
            class="port "
            data-name="out-0"
            data-nodeid="myId"
          >
            <div
              class="triangle-port"
            />
          </div>
        </div>
      </div>
      <div
        class="StatusLight"
      >
        <div
          class="StatusLightItem"
          style="background-color: red;"
        />
        <div
          class="StatusLightItem"
          style="background-color: white;"
        />
        <div
          class="StatusLightItem"
          style="background-color: white;"
        />
      </div>
      <div
        class="custom-node-description"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
